### PR TITLE
024 registration system

### PR DESCRIPTION
**Pull request by matthewknox99**

### CODE REVIEW

**Review by laura-puckoriute — APPROVED** [review]

File: src/main/java/org/kainos/ea/data/UserData.java [hunk 1]
```diff
@@ -65,4 +65,37 @@ public boolean removeToken( Connection conn, String email, String token ) throws
         return false;
 
     }
+
+    public String generateHash( String password ) {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/data/UserData.java]: this should be a private method so it doesn't require testing

File: src/main/java/org/kainos/ea/resources/WebService.java [hunk 2]
```diff
@@ -84,4 +89,27 @@ public Response getJobSpecification( @PathParam("id") int id ) throws SQLExcepti
...
+    public Response registerUser(UserRequest user) {
+        if (userValidator.isValidUser(user)) {
+            try {
+                int id = userService.registerUser(user);
+                return Response.status(HttpStatus.CREATED_201).entity(id).build();
+            } catch (DatabaseConnectionException | SQLException e) {
+                System.out.println(e);
+                return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500).build();
+            }
+            catch (DataNotFoundException e) {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/resources/WebService.java]: I don't see this exception being thrown anywhere, can you throw it when the roleID passed in does not exist in the db?

File: src/test/java/org/kainos/ea/integration/WebServiceIntegrationTest.java [hunk 3]
```diff
@@ -98,6 +104,43 @@ void getJobSpecification_shouldReturnErrorStatus404() throws UnsupportedEncoding
...
+                .post(Entity.entity(user, MediaType.APPLICATION_JSON_TYPE))
+                .getStatus();
+
+        assertEquals(expectedResult, response);
+    }
+
+    @Test
+    void registerUser_shouldThrow404Error_whenDataNotFoundExceptionThrown() {
+        int expectedResult = HttpStatus.NOT_FOUND_404;
+        Response response = APP.client().target( hostURI + registerEndpoint + "/test" )
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/WebServiceIntegrationTest.java]: this isn't a valid test, that endpoint doesn't exist

File: src/test/java/org/kainos/ea/service/UserServiceTest.java [hunk 4]
```diff
@@ -251,6 +257,47 @@ public void generateHash_shouldNotReturnIdenticalHash_whenGivenDifferentInput()
...
+        Mockito.when(databaseConnector.getConnection()).thenReturn(conn);
+        Mockito.when(userData.registerUser(userRequest, conn)).thenReturn(expectedResult);
+
+        int actualResult = userService.registerUser(userRequest);
+
+        assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    void registerUser_shouldThrow400Error_whenUserCredentialsAreInvalid() throws DatabaseConnectionException, SQLException {
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/UserServiceTest.java]: I'm assuming your userData class does not return a 400 if user credentials are invalid as that would be a valid ID

File: src/main/java/org/kainos/ea/validator/UserValidator.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package org.kainos.ea.validator;
+
+import org.kainos.ea.models.UserRequest;
+
+public class UserValidator {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/validator/UserValidator.java:6]: this needs unit tests

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/main/java/org/kainos/ea/data/UserData.java (hunk 1 above)
Comment (matthewknox99) [reply src/main/java/org/kainos/ea/data/UserData.java]: changed to private

File: src/main/java/org/kainos/ea/resources/WebService.java (hunk 2 above)
Comment (matthewknox99) [reply src/main/java/org/kainos/ea/resources/WebService.java]: exception thrown in UserService

File: src/test/java/org/kainos/ea/integration/WebServiceIntegrationTest.java (hunk 3 above)
Comment (matthewknox99) [reply src/test/java/org/kainos/ea/integration/WebServiceIntegrationTest.java]: removed

File: src/test/java/org/kainos/ea/service/UserServiceTest.java (hunk 4 above)
Comment (matthewknox99) [reply src/test/java/org/kainos/ea/service/UserServiceTest.java]: I'm not sure what you mean by this comment

File: src/main/java/org/kainos/ea/validator/UserValidator.java (hunk 5 above)
Comment (matthewknox99) [reply src/main/java/org/kainos/ea/validator/UserValidator.java:6]: added unit tests for valid/invalid email and password

File: src/test/java/org/kainos/ea/service/UserServiceTest.java (hunk 4 above)
Comment (matthewknox99) [reply src/test/java/org/kainos/ea/service/UserServiceTest.java]: test removed

**Review by shaunganley — APPROVED** [review]

**Pull request merged**